### ISSUE TITLE
Apparent increase in input latency with latest RA code on Windows

### ISSUE THREAD

**Issue by Ryunam** [libretro/RetroArch#9271]
## Description

I am noticing an apparent increase in input latency when testing the latest RA code, specifically with the latest self-compiled master through Mingw64 on Windows 10, and comparing its responsiveness with the latest code from the buildbot and with an older build from June.

Unfortunately I have little to no way to objectively measure this difference right now, but I performed a few tests using the Manual lag option of the **SNES 240p suite** using the latest snes9x core. 
The results, even if purely anecdotal, give a certain amount of credibility to the latency I felt while playing.

The following tests were performed with the same exact configuration, with Runahead enabled (!), same controller, same core, on the same device (a Surface Pro 6, same system I compiled RA on).

_**This is one first attempt with the latest code self-compiled through Mingw:**_
![240pSuite-190809-213740](https://user-images.githubusercontent.com/17614150/62804581-fdd32700-baed-11e9-83cf-6385dd4627b3.png)

_**This is a second attempt once again with the latest code self-compiled through Mingw:**_
![240pSuite-190809-214355](https://user-images.githubusercontent.com/17614150/62804898-d761bb80-baee-11e9-84e5-6d45ee96bc7a.png)

_**This is instead with the latest version available through the buildbot**_
![240pSuite-190809-213657](https://user-images.githubusercontent.com/17614150/62804686-48ed3a00-baee-11e9-8f8b-e6e6efafce86.png)

_**Finally, this is with the older June build**_
![240pSuite-190809-214156](https://user-images.githubusercontent.com/17614150/62804793-936eb680-baee-11e9-9fe8-109fc1a6219f.png)

I can reproduce this difference consistently by using either one of these three executables.

### Version/Commit
- RetroArch: 1.7.7 / latest code (e51b93993bdd556f8138f8ebcfab3105c2fc9bfc)

### Environment information
- OS: Windows 10 x64
- Compiler: Mingw64

**Comment by twinaphex** [comment]
If anything, it should be the other way around.

Anyway, we really need something more substantive than this. A bisect is really needed if you feel this is the case.

**Comment by twinaphex** [comment]
Maybe its the bitmask input code, although I would find this weird. Try to force the environment callback for it to false so that you can check if the non bitmask input state path yields better results.

Feel free to ask me how to do this in case its n ot clear.

**Comment by Ryunam** [comment]
Okay, I have taken the time to do a bisect while repeating the same test over and over and 4d860c4d16c09cc6b7e06966089abd92d0bcc1e5 appears to be the first bad commit.

The following anecdotal tests, performed with the same criteria described above (snes9x core, same exact configuration, same controller, same device), seem to back up my findings.

**_240p Suite test with RA on 4d860c4d16c09cc6b7e06966089abd92d0bcc1e5_**:
![240pSuite-190809-231621](https://user-images.githubusercontent.com/17614150/62809639-c53a4a00-bafb-11e9-89db-32d08b802f8b.png)

_**240p Suite test with latest RA code (commit 7c9de7ec0f40128f27586167506e2872051633fc)**_:
![240pSuite-190809-231858](https://user-images.githubusercontent.com/17614150/62809758-2530f080-bafc-11e9-9145-b7af040293af.png)

_**240p Suite test with RA on 6482a81ef5e294d282b38eea63de19425b0fe836 (1 commit before the bisected bad one)**_:
![240pSuite-190809-231143](https://user-images.githubusercontent.com/17614150/62809420-2f062400-bafb-11e9-974f-c0a992f255b0.png)

**Comment by Ryunam** [comment]
I am also attaching the two .exe files I've compiled, respectively the one built on commit 4d860c4d and the other one compiled with the commit right before (6482a81).
Perhaps others can test these two files with the "Manual Lag test" option of the 240p Suite and report their findings.

- [retroarch - before 4d860c4d.zip](https://github.com/libretro/RetroArch/files/3488095/retroarch.-.before.4d860c4d.zip)
- [retroarch - after 4d860c4d.zip](https://github.com/libretro/RetroArch/files/3488097/retroarch.-.after.4d860c4d.zip)

**Comment by twinaphex** [comment]
Hm, that is interesting, thanks.

**Comment by twinaphex** [comment]
Ok what happens if you take the most recent master/commit but just revert the bad commit? Does that fix it?

**Comment by twinaphex** [comment]
OK, try this instead -

https://github.com/libretro/RetroArch/commit/5cac6dddfb5f8039b256bf07b8ba0458ebed8b04

This is the only real change I could see. Basically, without the check to see if the menu driver is alive, I believe it could have rendered the libretro frame buffer a second time inadvertently. Let's see if this is what was the culprit and if this brings latency back down.

**Comment by twinaphex** [comment]
If this doesn't work, try to make a PR on top of master that solves the issue for you. I tried running 240ptestsuite myself but even going a git version back before the commit you point to as 'bad', I still get about 22ms latency with Manual Lag and sometimes a 4/5 here or there.

So I'd feel more comfortable if you instead would find a fix on your end that resolves the issue for you. I'd be fine with you reverting the culprit commit and leaving it at that, as long as it fixes the actual issue on your end and you can confirm it does.

**Comment by Ryunam** [comment]
@twinaphex, thank you as usual for your dedication and patience. i’ll test each one of your suggestions in about a hour and will report back here for sure.

**Comment by Ryunam** [comment]
Okay, I have performed more tests with the same method and here are the results. 

I am happy to confirm that the latest master (562ea776133e328651d500081a7eef33ab68c957) seems to fix the regression I was experiencing on my end and brings latency performance back to how it was before.

(Actually it could be even better now? It would have to be measured more objectively though)

-----

_**240p Suite with latest master (562ea776133e328651d500081a7eef33ab68c957)**_:

![240pSuite-190810-113126](https://user-images.githubusercontent.com/17614150/62820290-7f649c80-bb62-11e9-85d8-026f316c73ab.png)

_**240p Suite compiled on commit 4d860c4d16c09cc6b7e06966089abd92d0bcc1e5**_:

![240pSuite-190810-113310](https://user-images.githubusercontent.com/17614150/62820308-b5a21c00-bb62-11e9-9273-41cd1400c75d.png)

-----

I have also tried to do a bisect in order to understand which commit solved the issue and there is a clear difference before and after 5cac6dddfb5f8039b256bf07b8ba0458ebed8b04.

_**240p Suite before 5cac6dddfb5f8039b256bf07b8ba0458ebed8b04**_:

![240pSuite-190810-114101](https://user-images.githubusercontent.com/17614150/62820439-9efcc480-bb64-11e9-8430-3d2011404dba.png)

_**240p Suite after 5cac6dddfb5f8039b256bf07b8ba0458ebed8b04**_:

![240pSuite-190810-114546](https://user-images.githubusercontent.com/17614150/62820452-c05db080-bb64-11e9-9219-2f4b568a75e5.png)

-----

Based on all of the above, I would consider this issue to be fixed. Thank you once again @twinaphex for quickly implementing a fix.

That being said I would also like to leave this issue open for just a few more days, in case anyone else wants to chime in and contribute with their own findings and in order to make sure that the latency regression is fully addressed for everyone.
